### DISCUSSION TITLE
[feature request] Opening files from a "git status" list view

### DISCUSSION THREAD

**Discussion by dikarel** [zed-industries/zed#14099] [Feature Requests]
I'm exploring Zed as a faster alternative to VSCode -- I don't mind losing the VSCode bells and whistles to get a snappier UI that can handle a larger complex monorepo codebase without stuttering.

So far so good, but a potential deal-breaker for me is the lack of a "list of files" view that's based on `git status`. The default "project browser" view is a good table stakes to have, but is not very usable in a corporate monorepo situation with 1000s of files located across many directories. For a typical engineering task, I'd need to open 10+ files across disparate locations in the monorepo. Using a regular project tree UI, it's hard to switch between relevant files and get a holistic view of where I'm at with regards to progress.

Here's what the requested feature looks like in VSCode (filenames censored). I usually use this view as a way to keep track and jump quickly between files that I'm editing. I don't need anything fancy here -- just having a read-only list of "git status" output would be a quantum leap from the status quo.

<img width="330" alt="Screenshot 2024-07-10 at 1 31 46 PM" src="https://github.com/zed-industries/zed/assets/15131271/f2832e76-3fec-4ca8-9007-b9fedaba4922">

Right now, my workaround is to `git status` in a separate terminal and manually copy-paste the paths into Zed's fuzzy search feature -- very inefficient and flow-breaking.

PS: Apologies if this feature already exists -- at the moment it's unclear how I might reach it if so. I also don't mind being told to "please contribute an extension" if that's actually possible.

**Comment by SomeoneToIgnore** [comment]
There's a draft branch that has a first approximation of the feature: https://github.com/zed-industries/zed/issues/8665#issuecomment-2194000497

Contrasting to the VSCode impl, this one will have a project search-like multi buffer with all git changes + the outline panel to show the git-related entries.

The outline panel with filters is there, but there's no way to show anything specific (search results, git entries, diagnostics, etc.), only the outlines.
Fixing this would be the next good step, and the prototype branch needs more polishing which is not done yet.

Overall, this is a long road and getting there might take up to the end of this year.

**Reply by dikarel** [reply]
Thanks! It is indeed a dupe of a few existing asks. Maybe a thing worth mentioning is that I believe _just shipping a read-only list view is sufficient_. No need for the other bells and whistle features [1] to un-block adoption of this IDE as a daily driver -- just that read-only feature alone is a pretty big deal for me.

[1] e.g. git commit, git fetch/pull, diff-n-merge tools

**Reply by SomeoneToIgnore** [reply]
For now, there's a way to call a Zed task https://zed.dev/docs/tasks with specific context (e.g. project root path) to call an external diff tool which might help somewhat.

Read-only impl might take similar amount of time to implement, as the main current issues are
* Zed's blindness towards conflicted/staged/deleted/added files — it's just inner file changes that are shown with hunks now
* a way to update due to external and internal events — as even the readonly view has to update if a neighbour buffer got an edit, or some git command was run in the terminal — that is covered by the PR mainly, but some rough edges persist

**Reply by dikarel** [reply]
Thanks for sharing technical details!

To help unblock others who might land in this ticket, I found a usable temp workaround which is to open a terminal panel and run a Bash script to poll `git status`:

`while true; do clear; git status --porcelain; sleep 5; done`

Since I can command+click into the filenames to open them, that's good enough to unblock my use-case.

<img width="1070" alt="shapes at 24-07-10 16 05 19" src="https://github.com/zed-industries/zed/assets/15131271/eb50b447-64f2-4c45-b337-6a1ede123783">
